airbyte-integrations/connectors/source-mailchimp/source_mailchimp/source.py

```diff
@@ -21,70 +21,222 @@
+import math
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple
+
+import requests
+from airbyte_cdk import AirbyteLogger
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator
+from mailchimp3 import MailChimp
+
+
+class MailChimpStream(HttpStream, ABC):
+    url_base = "https://us2.api.mailchimp.com/3.0/"
+    primary_key = "id"
+    PAGINATION = 100
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # self.url_base = url_base #but `us2` can differ as {ds} in docs
+        self.current_offset = 0
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        decoded_response = response.json()
+        api_data = decoded_response[self.data_field]
+        if len(api_data) < self.PAGINATION:
+            self.current_offset = 0
+            return {}
+        else:
+            self.current_offset += self.PAGINATION
+            return {"offset": self.current_offset}
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+
+        params = {"count": self.PAGINATION}
+
+        # Handle pagination by inserting the next page's token in the request parameters
+        if next_page_token:
+            params.update(next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        yield from response_json[self.data_field]
+
+    @property
+    @abstractmethod
+    def data_field(self) -> str:
+        """the responce entry that contains useful data"""
+        pass
+
+
+class IncrementalMailChimpStream(MailChimpStream, ABC):
+    state_checkpoint_interval = math.inf
+
+    @property
+    @abstractmethod
+    def cursor_field(self) -> str:
+        """
+        Defining a cursor field indicates that a stream is incremental, so any incremental stream must extend this class
+        and define a cursor field.
+        """
+        pass
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+
+        current_state = current_stream_state.get(self.cursor_field)
+        latest_state = latest_record.get(self.cursor_field)
+
+        if not all([current_state, latest_state]):
+            nonzero_state = list(filter(bool, [current_state, latest_state]))[0]
```

Comment:
is there is a change that we will have a null value in the cursor?

Reply:
First time when we load the state (from very begining), one of states is empty and we get null (checked that). But we always have one not-null state (checked that also printing states to log) - both null states never take a place.